### ISSUE TITLE
Rofi auto focus not work after upgrade to hyprland version 0.37.1

### ISSUE THREAD

**Issue by hesam-init** [hyprwm/Hyprland#5142]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch HEAD at commit c5e28ebcfe00a510922779b2c568cfa52a317445  (props: bump ver 0.37.1).
Date: Sat Mar 16 14:51:49 2024
Tag: v0.37.1

flags: (if any)


System Information:
System name: Linux
Node name: void
Release: 6.6.22_1
Version: #1 SMP PREEMPT_DYNAMIC Fri Mar 15 23:56:15 UTC 2024


GPU information:
01:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Ellesmere [Radeon RX 470/480/570/570X/580/580X/590] [1002:67df] (rev e7) (prog-if 00 [VGA controller])


os-release: NAME="Void"
ID="void"
PRETTY_NAME="Void Linux"
HOME_URL="https://voidlinux.org/"
DOCUMENTATION_URL="https://docs.voidlinux.org/"
LOGO="void-logo"
ANSI_COLOR="0;38;2;71;128;97"

DISTRIB_ID="void"


plugins:
```

</details>

### Bug or Regression?

Bug

### Description

Hey i newly upgraded my hyprland version to 0.37.1 and eveything works perfectly even the steam freeze bug fixed at this version but rofi menu will not auto focus on launch until i hover my cursor on it. here is this video :

https://github.com/hyprwm/Hyprland/assets/89181937/c48025b1-0c48-4ad2-8330-084ae9c3e7f0

os : void linux

i installed all of required packages for hyprland : 
hyprland-protocols xorg-server-xwayland xdg-desktop-portal-hyprland xdg-desktop-portal xdg-utils wayland wayland-protocols xdg-desktop-portal-wlr xdg-desktop-portal-gtk 

### How to reproduce

my hyprland configs : 

[.dots-hyprland](https://github.com/hesam-init/.dots-hyprland)

**Comment by M4he** [comment]
Have you tried [the wayland fork of rofi](https://github.com/lbonn/rofi)? The fork never had issues on Hyprland for me.

I just tried plain rofi out of curiosity. It doesn't use fractional scale, has blur rendering artifacts at borders and corners and has the focus issue you described. I'd strongly recommend checking out the fork.

**Comment by vaxerski** [comment]
won't be fixed because it's xwayland shenanigans

**Comment by JadeJitsu** [comment]
> Have you tried [the wayland fork of rofi](https://github.com/lbonn/rofi)? The fork never had issues on Hyprland for me.
> 
> I just tried plain rofi out of curiosity. It doesn't use fractional scale, has blur rendering artifacts at borders and corners and has the focus issue you described. I'd strongly recommend checking out the fork.

I had the same issue and confirm that moving to the Wayland fork of Rofi fixed the issue.

**Comment by mylinuxforwork** [comment]
Fixed it with rofi by adding the windowrule

`windowrulev2 = stayfocused,class:(Rofi)`

**Comment by mylinuxforwork** [comment]
Not fixed. Window is focused but not activated for keyboard input. Adding 

`windowrulev2 = forceinput,class:(Rofi)`

brings no change. @vaxerski  any idea?
The rofi wayland fork brings more side effects with waybar. so want to stick with rofi.

**Comment by vaxerski** [comment]
hyprland will refuse focus to OR surfaces that don't explicitly allow it. Input sniffing (like in rofi's case) is exploiting Xorg and won't be supported.

**Comment by mylinuxforwork** [comment]
Thanks for your feedback.

**Comment by addz** [comment]
Thanks. Hit this issue as well after upgrading to latest Hyprland package in arch repo (`0.37.1-1`, not using aur's git version).

Moved to the lbonn fork, works great, looks slightly better, and feels snappier (I've been aware that it's running on x11 but been too lazy to replace it)

@mylinuxforwork - not sure what waybar side effects you're getting - but maybe it's time to fix those instead? If you're running a Wayland setup, it deserves to have it's launcher native Wayland too :stuck_out_tongue_winking_eye:



(btw thanks @vaxerski - love Hyprland and all the good work you're putting out :metal: )

**Comment by mylinuxforwork** [comment]
I moved to Wayland rofi as well and fixed the issues. Could solve all waybar sideeffects in my dotfiles.

**Comment by Sewer56** [comment]
If you're using extensions like `rofi-emoji`, then you might need also to rebuild them if switching to a fork of `rofi`.
In my case, using `rofi-emoji-git` did the trick.

**Comment by mylinuxforwork** [comment]
Thanks for your feedback.

**Comment by Atemo-C** [comment]
> A quick fix if someone needs it. You can manually set focus to the rofi window after a short delay with hyprctl dispatcher.
> 
> Example: `bind = [superkey], [key], exec, [rofi script path or command] & sleep 0.2; hyprctl dispatch focuswindow "^(Rofi)"`

As I wait for the few X-only menu programs I use to be updated to support Wayland, this does the job of dealing with those Xwayland shenanigans. This is a bit hacky and adds a bit of a delay before the program gets focused, but it's better than having to manually focus it each time. Thanks!

*Also note that the delay `0.2` might be changed to be higher if you have a slower drive/the program takes longer to launch on your system.*

**Comment by Trinxic** [comment]
Instead of creating a delay, you can just use the `focuswindow` disbatcher:

`<rofi_keybind>, focus, window, title:(rofi), floating`

I'm also using this window rule: `windowrulev2 = stayfocused, class:(Rofi)`

This seemed to work for me

**Comment by Atemo-C** [comment]
> Instead of creating a delay, you can just use the `focuswindow` disbatcher:
> 
> `<rofi_keybind>, focus, window, title:(rofi), floating`
> 
> I'm also using this window rule: `windowrulev2 = stayfocused, class:(Rofi)`
> 
> This seemed to work for me

This does not seem to work properly for me without a delay; My guess is that I implemented this improperly *(I am still not that used to Hyprland).*
What does the line with the bind in question look like in your configuration?

**Comment by CatskinnerTech** [comment]
Ok, getting rid of the second keybinding, just use @Atemo-C 's `windowrulev2`

**Comment by Pikatsuto** [comment]
Hello, I take the liberty of reopening this post just long enough to post a solution that should work for everyone, because the previous ones don't work for me either.

```
# Se but waiting for rofi to open then trigger focus
$focusRofi = & while [ "$(hyprctl clients | grep "class: Rofi")x" == "x" ]; do continue; done; hyprctl dispatch focuswindow "^(Rofi)"

# Add variable behind each rofi execution
bind = $mainMod, D, exec, rofi -show drun $focusRofi
bind = $mainMod, M, exec, rofi-beats $focusRofi
bind = $mainMod, V, exec, rofi-vm $focusRofi
bind = $mainMod, W, exec, rofi-wpa $focusRofi
bind = $mainMod, B, exec, rofi-bluetooth $focusRofi
```

Sorry for the English, I'm using a translator.

**Comment by maxolasersquad** [comment]
This is fixed in Rofi 2.0, which now  has first-class support for Wayland. I have compiled it on my Ubuntu 24.04 machine and can confirm that it now has focus when I use my keyboard shortcut to launch it (`$mainMod, R` for me).
